### DISCUSSION TITLE
https://maxkim-j.github.io/posts/nuxt-ssr

### DISCUSSION THREAD

**Discussion by MaxKim-J** [MaxKim-J/maxkim-j.github.io#85] [Announcements]
#          🌧 Nuxt로 살펴보는 서버사이드 렌더링의 핵심      

https://maxkim-j.github.io/posts/nuxt-ssr

**Comment by MaxKim-J** [comment]
<p>진짜 딥하게 정리 잘하셨네요! 덕분에 잘 읽고 갑니다.<br>2가지 질문을 드리고 싶습니다!<br>URL 요청해서 SSR 이 돌아서 컴포넌트가 렌더링이 된거는 이해했습니다.<br>그 이후에서 사이트에서 링크 이동하면 CSR 처럼 동작하는거도 알고는 있습니다.</p><p>그런데 SSR 이 발생하지 않는데, asyncData() 훅같은 생명주기 메서드들이 어떻게 실행이 되는지 알고싶습니다.(힌트라도 주시면 감사하겠습니다)</p><p>그리고 Nuxt.js 서버는 어떤 책임을 가지고 있다고 생각하시나요?<br>당연히 SSR 및 SPA 단점을 보완하기 위한 책임까지다라고도 이야기할수도 있겠고<br>아니다. 잘만 만들면 MPA 와 유사하게 만들수도있다 이럴수도있는데 글쓴이님 생각이 궁금합니다.</p><p>저같은 경우는 Nuxt 에 서버를 가끔 스프링과 같은 서버처럼 생각해서 API 도 거기서 다 만들면 되는거 아니야? 이러시는 분들이 있는데 ㅠㅠ<br>Nuxt 같은 경우는 말 그대로 Vue 를 기반으로 SPA 를 좀 더 쉽고 간편하게 개발을 하게 해주는 기술 + SPA 에 단점을 보완해주는 책임만 있으면 된다고 생각하는데 ㅠㅠ</p>

>_This comment was made by Disqus. 2020-11-26 11:35:29 라이언(disqus_Bam85l2bkM)_

**Comment by MaxKim-J** [comment]
<p>안녕하세요!! 포스트 잘 읽어주셔서 감사합니다.</p><p>먼저 asyncData는 실험을 해보니 라이언님 말씀처럼 서버와 클라이언트에서 모두 호출될 수 있는 훅이네요(본문에 서버에서만 호출되는 훅으로 쓰여져서 본문 수정하겠습니다..!) 근데 실험 결과 재미있는 점이 asyncData()가 서버에서 호출되었다면 클라이언트에서 호출되지 않는 훅이기도 합니다(SSR이건 CSR이건 한번만 호출되네요) 정확하진 않지만 CSR과정에서 Vue 인스턴스가 만들어지기 이전에 자바스크립트가 asyncData()의 호출 여부를 타진하고 컴포넌트 렌더링 전에 무조건 한번 실행하는게 아닌가 싶습니다. Vue 생명주기 훅이 아닌 Nuxt만의 훅들은 라이브러리의 코어 단에서 미들웨어라던지, 다른 생명주기 훅을 호출하는 자바스크립트 코드를 붙여서 빌드하는게 아닌가 생각해봤는데... 사실 Nuxt 코드를 좀 더 뜯어봐야 정확한 동작을 알 수 있을 것 같습니다 부족한 답변 죄송해요ㅜㅜ</p><p>저도 라이언님처럼 Nuxt의 효용을 'SPA의 단점을 보완하기 위해 SSR을 제공한다는 점', '라우팅이나 플러그인 라이브러리들을 간편하게 사용할 수 있는 인터페이스를 제공한다는 점' 딱 두 가지로 생각하고 있습니다. 물론 서버 로직을 커스텀하면 API나 MPA를 개발할 수 있겠지만 Nuxt 공식 문서에서 그렇게 개발할때도 사용할 수 있다는 지침은 찾기 힘들어 확실히 그러라고 만든 라이브러리는 아닌 것 같습니다. 그리고 정말 여의치 않은 경우가 아니면 하나의 배포 단위는 일관성 있는 인터페이스를 제공해야한다고 생각하는 쪽이라 요청시 응답에 REST와 SPA가 섞여있는 구조가 좋은 구조라고 말하기도 힘들고요. 물론 리액트 진영의 Next는 공식문서에 express등을 사용해 API를 개발할 수 있다고 명시해놓긴 했습니다만, 섞여있는 것 보다는 하나의 목적을 바탕으로 개발하는 것이 좋지 않을까 하는 생각입니다.</p>

>_This comment was made by Disqus. 2020-11-28 11:38:46 Max Kim(disqus_O30MHz2G2q)_

**Comment by MaxKim-J** [comment]
<p>친절한 답변 너무 감사합니다!<br>코로나 조심하시고 업계에서 멋진 프론트엔드 개발자로 만나길 기원할게요!<br>좋은 주말되세요 ^^</p>

>_This comment was made by Disqus. 2020-11-28 11:58:04 라이언(disqus_Bam85l2bkM)_

**Comment by MaxKim-J** [comment]
<p>오랜만에 놀러옴.<br>ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ</p>

>_This comment was made by Disqus. 2021-04-28 07:39:40 라이언(disqus_Bam85l2bkM)_

**Comment by MaxKim-J** [comment]
<p>”불완전한 HTML 파일이라는 ‘마른 땅’에 자바스크립트라는 ‘물’을 뿌리는 일”<br>갬동쓰..</p>

>_This comment was made by Disqus. 2022-01-03 10:56:57 sulmo J(sulmoj)_

**Comment by ycy3723** [comment]
겉만 핥는 5만원짜리 강의 사서 시간낭비, 돈낭비 하는 것 보다 이 글이 몇배는 훌륭한 것 같습니다. 제가 하루종일 검색하고 문서 읽으면서도 간지러운 부분들도 많고, 막연한 개념들이 정리가 안되어 있었는데 이 글을 읽고 정리됐습니다. 작성자님 정말 감사합니다🙇

**Comment by geunheejung** [comment]
좋은 글 감사합니다!
